controller/api/util/api_utils.go

```diff
@@ -231,3 +245,176 @@ func buildResource(namespace string, resType string, name string) (pb.Resource,
 		Name:      name,
 	}, nil
 }
+
+func BuildTapByResourceRequest(params TapRequestParams) (*pb.TapByResourceRequest, error) {
+	target, err := BuildResource(params.Namespace, params.Resource)
+	if err != nil {
+		return nil, fmt.Errorf("target resource invalid: %s", err)
+	}
+	if !contains(ValidTargets, target.Type) {
+		return nil, fmt.Errorf("unsupported resource type [%s]", target.Type)
+	}
+
+	matches := []*pb.TapByResourceRequest_Match{}
+
+	if params.ToResource != "" {
+		destination, err := BuildResource(params.ToNamespace, params.ToResource)
+		if err != nil {
+			return nil, fmt.Errorf("destination resource invalid: %s", err)
+		}
+		if !contains(ValidDestinations, destination.Type) {
+			return nil, fmt.Errorf("unsupported resource type [%s]", target.Type)
+		}
+
+		match := pb.TapByResourceRequest_Match{
+			Match: &pb.TapByResourceRequest_Match_Destinations{
+				Destinations: &pb.ResourceSelection{
+					Resource: &destination,
+				},
+			},
+		}
+		matches = append(matches, &match)
+	}
+
+	if params.Scheme != "" {
+		match := buildMatchHTTP(&pb.TapByResourceRequest_Match_Http{
+			Match: &pb.TapByResourceRequest_Match_Http_Scheme{Scheme: params.Scheme},
+		})
+		matches = append(matches, &match)
+	}
+	if params.Method != "" {
+		match := buildMatchHTTP(&pb.TapByResourceRequest_Match_Http{
+			Match: &pb.TapByResourceRequest_Match_Http_Method{Method: params.Method},
+		})
+		matches = append(matches, &match)
+	}
+	if params.Authority != "" {
+		match := buildMatchHTTP(&pb.TapByResourceRequest_Match_Http{
+			Match: &pb.TapByResourceRequest_Match_Http_Authority{Authority: params.Authority},
+		})
+		matches = append(matches, &match)
+	}
+	if params.Path != "" {
+		match := buildMatchHTTP(&pb.TapByResourceRequest_Match_Http{
+			Match: &pb.TapByResourceRequest_Match_Http_Path{Path: params.Path},
+		})
+		matches = append(matches, &match)
+	}
+
+	return &pb.TapByResourceRequest{
+		Target: &pb.ResourceSelection{
+			Resource: &target,
+		},
+		MaxRps: params.MaxRps,
+		Match: &pb.TapByResourceRequest_Match{
+			Match: &pb.TapByResourceRequest_Match_All{
+				All: &pb.TapByResourceRequest_Match_Seq{
+					Matches: matches,
+				},
+			},
+		},
+	}, nil
+}
+
+func buildMatchHTTP(match *pb.TapByResourceRequest_Match_Http) pb.TapByResourceRequest_Match {
+	return pb.TapByResourceRequest_Match{
+		Match: &pb.TapByResourceRequest_Match_Http_{
+			Http: match,
+		},
+	}
+}
+
+func contains(list []string, s string) bool {
+	for _, elem := range list {
+		if s == elem {
+			return true
+		}
+	}
+	return false
+}
+
+func RenderTapEvent(event *pb.TapEvent) string {
```

Comment:
This feels like a CLI specific method? I had assumed that web would pass back a struct we could slice and dice on the frontend side of things.

Reply:
yes! that is what I was referring to when I said this in the PR description:
>  I've moved a bunch of code from the cli tap.go into utils so that the code can be shared between web and CLI. I think we should consider making the display more suited to web, but in the short term, reusing the CLI's rendering of tap events works.

Reply:
👍 RTFM here I come!